plugins/npm/npm.plugin.zsh

```diff
@@ -1,16 +1,10 @@
-(( $+commands[npm] )) && {
-    __NPM_COMPLETION_FILE="${ZSH_CACHE_DIR}/npm_completion"
-
-    if [[ ! -f $__NPM_COMPLETION_FILE ]]; then
-        npm completion >! $__NPM_COMPLETION_FILE 2>/dev/null
-        [[ $? -ne 0 ]] && rm -f $__NPM_COMPLETION_FILE
-    fi
-
-    [[ -f $__NPM_COMPLETION_FILE ]] && source $__NPM_COMPLETION_FILE
-
-    unset __NPM_COMPLETION_FILE
-}
-
+(( $+commands[npm] )) &&
+(
+  COMPLETION="${ZSH_CACHE_DIR}npm_completion"
+  [[ ! -f $COMPLETION ]] && npm completion 2>/dev/null >! $COMPLETION
+  source $COMPLETION
+  npm completion 2>/dev/null >! $COMPLETION &
```

Comment:
It will cache npm completion on first run. And will show you wrong completion if you will switch to older/newer version of npm (for example with [n](https://github.com/tj/n))